Always add new TaskGroup to TaskPrefix

jrbourbeau opened the pull request:
Currently a `TaskGroup` is only added to a `TaskPrefix` when a new `TaskPrefix` is created:

https://github.com/dask/distributed/blob/d747f639d73c5f9b8961f2f80e407868de7da4cb/distributed/scheduler.py#L1969

I think we want to add append a `TaskGroup` to `TaskPrefix.groups` anytime a new group is created. 

Closes #3368

mrocklin commented:
Thanks for catchiing this @jrbourbeau !  In principle this looks good to me, and I'm always happy to see people engaging on the scheduler.  It looks like there is a failing test here.

jrbourbeau commented:
Hmm the test appears to be failing due how we handle tasks like `"inc-123"`. These types of tasks belong to the `TaskPrefix` I expected (task `"inc-123"` belongs to `TaskPrefix` `"inc"`), however they all belong to the same `TaskGroup` with an empty string name:

```python
In [1]: from distributed.utils import key_split_group, key_split

In [2]: key_split("inc-123")    # task prefix name
Out[2]: 'inc'

In [3]: key_split_group("inc-123")    # task group name
Out[3]: ''

In [4]: key_split_group("add-456")    # task group name
Out[4]: ''
```

I _think_ what we want (certainly open to other suggestions) is to have a separate group for keys like `"inc-123"` such that `"inc-123"`, `"inc-456"`, etc. belong to an `"inc"` `TaskGroup` instead of the `""` `TaskGroup`. So something like:

`"inc-123"` -> `TaskGroup("inc")` -> `TaskPrefix("inc")`
`"inc-456"` -> `TaskGroup("inc")` -> `TaskPrefix("inc")`
`"(inc-789, 0)"` -> `TaskGroup("inc-789")` -> `TaskPrefix("inc")`

It's worth noting that this will result in an increased number of task groups compared to our current behavior

mrocklin commented:
So, *currently* task groups track dependencies, and are acyclic.  Having all `inc-123` tasks go to the `inc` task group would break this.  Today this doesn't matter (we haven't built up much around TaskGroups yet) but I imagine that it might in the near future.

Having inc-123 go to a task group of inc-123 would mean that we have *many* task groups in dask.delayed or futures workloads.  We currently let TaskGroups persist after all of their tasks are gone.

mrocklin commented:
The solution so far has been to just not give tasks like `inc-123` a TaskGroup (or assign them the `None` group) which I think makes some sense.  They aren't really part of a group of tasks.

jrbourbeau commented:
Thanks for the feedback!

The main reason I proposed having tasks like `"inc-123"` belong to a group `TaskGroup("inc")` was I was confused about how tasks are currently assigned a `TaskPrefix` and `TaskGroup`. For example, task `"inc-124"` is tracked by `TaskGroup("")` and `TaskPrefix("inc")`. But it wasn't clear to me that `TaskGroup("")` should belong to `TaskPrefix("inc")`'s `.groups` list. There can be many other tasks, e.g. `"add-456"`, which also belong to `TaskGroup("")` but have a different `TaskPrefix`, e.g. `TaskPrefix("add")`. This confusion may come from a lack of understanding on my part. 

As an alternative, we could continue to assign tasks like `"inc-123"` to `TaskGroup("")`, but also have them belong to a null task prefix like `TaskPrefix("")` that only tracks the `TaskGroup("")` group?

> So, currently task groups track dependencies, and are acyclic

Thanks for pointing that out. I wasn't thinking about how task group dependencies fit into the proposed changes. FWIW, `TaskGroup("")` can have itself as a dependency because tasks like `inc-123` and `add-456` both belong to that group and can depend on one another:

<details>
<summary>Example:</summary>

```python

from distributed import Client, wait

if __name__ == "__main__":
    
    inc = lambda x: x + 1
    add = lambda x, y: x + y
    
    with Client() as client:
        f_inc = client.submit(inc, 1)
        f_add = client.submit(add, f_inc, 2)
        wait([f_inc, f_add])
        
        group = client.cluster.scheduler.task_groups[""]
        print(f"group = {group}")
        print(f"group in group.dependencies = {group in group.dependencies}")
```

outputs

```
group = <no-group: memory: 2>
group in group.dependencies = True
```

</details>

mrocklin commented:
> I was confused about how tasks are currently assigned a TaskPrefix and TaskGroup

Yeah, this is all fairly new and rough.  I'm sure that we'll find better ways as we continue down this path.

> But it wasn't clear to me that TaskGroup("") should belong to TaskPrefix("inc")'s .groups list.

This is not clear to me either

> There can be many other tasks, e.g. "add-456", which also belong to TaskGroup("") but have a different TaskPrefix, e.g. TaskPrefix("add"). This confusion may come from a lack of understanding on my part

Yup.  Same.

> FWIW, TaskGroup("") can have itself as a dependency because tasks like inc-123 and add-456 both belong to that group and can depend on one another:

Yes, I'm mostly planning to ignore that group in future work.

> As an alternative, we could continue to assign tasks like "inc-123" to TaskGroup(""), but also have them belong to a null task prefix like TaskPrefix("") that only tracks the TaskGroup("") group?

Currently the prefix is used for the `duration_average` attribute, which affects scheduling decisions.  It's important for `inc-123` to map to the TaskPrefix `inc`

jrbourbeau commented:
I've updated this PR to remove `TaskGroup`s once all their tasks are in the `"forgotten"` state and also included a modified version of the test added over in https://github.com/dask/distributed/pull/3370 (thanks @quasiben!)

To reiterate the main changes in this PR are:

- Tasks with keys like `"inc-123"` (e.g. from `client.submit(inc, 1)`) are now added to `TaskGroup("inc")`. Previously they were added to a single `TaskGroup("")`.

- Because we are creating more task groups, we now remove task groups once all their tasks are in the `"forgotten"` state. This is to avoid holding onto many task groups in `dask.delayed` or futures workloads.

quasiben commented:
Looks great @jrbourbeau  -- thanks for taking this up

bnaul commented:
this seems like a fairly innocuous change, @jrbourbeau @jacobtomlinson any reason to hold off on merging? just don't want it to miss the next release since IMO the dashboard is currently close to useless for a complex workflow with many different tasks

jrbourbeau commented:
Thanks for following up @bnaul. Let's give a little more time for an additional review. I agree this should be included in the next bi-weekly release to fix known dashboard issues

mrocklin commented:
If you all are comfortable with this then please don't wait up for me

jrbourbeau commented:
Thanks @quasiben and @jacobtomlinson for reviewing!
